src/asm_parse.cpp

```diff
@@ -182,8 +191,20 @@ Instruction parse_instruction(const std::string& line, const std::map<std::strin
             .is_load = false,
         };
     }
-    if (regex_match(text, m, regex("lock " DEREF PAREN(REG PLUSMINUS IMM) " [+]= " REG))) {
-        return LockAdd{.access = deref(m[1], m[2], m[3], m[4]), .valreg = reg(m[5])};
+    if (regex_match(text, m, regex("lock " DEREF PAREN(REG PLUSMINUS IMM) " " ATOMICOP " " REG))) {
+        Atomic::Op op = str_to_atomicop.at(m[5]);
+        return Atomic{
+            .op = op,
+            .fetch = (op == Atomic::Op::XCHG || op == Atomic::Op::CMPXCHG),
+            .access = deref(m[1], m[2], m[3], m[4]),
+            .valreg = reg(m[6])};
+    }
+    if (regex_match(text, m, regex("lock " DEREF PAREN(REG PLUSMINUS IMM) " " ATOMICOP " " REG " fetch"))) {
+        return Atomic{
+            .op = str_to_atomicop.at(m[5]),
+            .fetch = true,
+            .access = deref(m[1], m[2], m[3], m[4]),
+            .valreg = reg(m[6])};
```

Comment:
I think it can be merged to a single branch with `"( fetch)"` being at 7 and checked for `.matched`)

Reply:
Done